### PR TITLE
Do better bindings

### PR DESCRIPTION
**Pull request by Odrakir**
## What's this?
This is just a proposal, not to be merged as it is, because if you like it it still needs some more code to apply these changes to the rest of the extensions.

Basically it's a way to improve the syntax on how we do bindings. The current way of doing this doesn't look really nice. Basically because of the two `Bind`... it's a little weird.

```c#
this.Bind(DataContext.CtaButtonTitle, CtaButton.BindTitle());
```

I propose we expose reactive properties of UIKit via an `Rx()` method (I wish it could be a property) in compatible types.

For example, if we want to access the tap observable of a button we would do:

```c#
CtaButton.Rx().Tap()
     .Subscribe(...)
```

This helps with discoverability of reactive properties, as they all will be under `Rx()` when code completing.

And also transforms the previous binding thing into:

```c#
this.Bind(DataContext.CtaButtonTitle, CtaButton.Rx().Title());
```

Which is nicer in my opinion.

## Why do we want this?
Cleaner code, easier to write (auto complete). Also we could continue this hostile takeover of the app by RxSwift (@kornelijepetak dixit) and add a thing called `ControlProperty` which would allow us to bind some properties as observables or observers depending on the call. For example, the text on a `UITextField` should be able to be both depending on the needs.

## How is it done?
The main thing is in the `Reactive.cs` file. The rest are extensions to that class changing the `where` clause to the appropriate UIKit component. I changed some in `UIButton` to prove my code.

## :squid: Permissions
Don't.

### CODE REVIEW

File: Toggl.Daneel/Extensions/UIKitRxExtensions/Reactive.cs [hunk 1]
```diff
@@ -0,0 +1,20 @@
+﻿namespace Toggl.Daneel.Extensions
+{
+    public class Reactive<TBase>
+    {
+        public TBase Base { get; set; }
+
+        public Reactive(TBase _base)
```
[the remark this answers is not in the document]
Comment (amulware) [reply Toggl.Daneel/Extensions/UIKitRxExtensions/Reactive.cs]: I thought you were gonna say "build aqueducts". ➡️ 🚪

**Comment by khoi** [comment]
💯 💯 💯

**Comment by kornelijepetak** [comment]
Since I'll be away for a few weeks, when my comments are heeded and acted upon, I hereby allow merge of this PR without my explicit approval.

**Comment by Odrakir** [comment]
I think this can wait until I'm back. Unless Williams wants it so bad he decides to do it himself.

**Comment by khoi** [comment]
I'm moving all UIKit Rx extensions to use the new `Reactive` class. I'm running in to 1 problem:

```
public static IObservable<string> Text<T>(this Reactive<T> reactive) where T: UITextField
public static IObservable<string> Text<T>(this Reactive<T> reactive) where T: UITextView
```

I can't declare those functions above since they share the same method signature. I can think of 2 solutions for it:

1. name the function differently for each type
1. 🔥 the generic constraint using `where`

```
public static IObservable<string> Text(this Reactive<UITextField> reactive) 
public static IObservable<string> Text(this Reactive<UITextView> reactive) 
```

the 2nd option has 1 major drawback
```
public static IObservable<string> IsEnabled(this Reactive<UIControl> reactive) 
public static IObservable<string> Text(this Reactive<UITextField> reactive) 

someTextField.Rx().IsEnabled // doens't have access 🙅‍♂️ 
```

Any thoughts?

**Comment by amulware** [comment]
You might be able to make the second option work out if the Reactive type is changed to be covariant in T (`<out T>`). Then the cast from Reactive<Subclass> to Reactive<Baseclass> should work...

**Comment by amulware** [comment]
Maybe you'd need to use an `IReactive<out T>` interface, in which case you can hide the actual reactive class inside the static extension class that creates it.

**Comment by daividssilverio** [comment]
🤔 `UITextField` and `UITextView ` share a common implementations `IUITextInputTraits`, `IUITextInput` for example. You could test:
```
public static IObservable<string> Text<T>(this Reactive<T> reactive) where T: <Insert Common Implementation Here>
```
instead of the specific ones.

**Comment by khoi** [comment]
>UITextField and UITextView share a common implementations IUITextInputTraits, IUITextInput for example. 
That doesn't work since UITextInputTraits doesn't expose the property we want, in this case `Text`

I tried to use covariant but wasn't successful. All of the classes are ported to the new binding method. 

The tricky one left is `UITextViewRxExtensions.cs` where the method would collide. I need help from some C# masters here :D

**Comment by willsb** [comment]
This can now be closed as #3236 was merged

**Pull request closed without merging**